ambry-frontend/src/main/java/com.github.ambry.frontend/AccountAndContainerInjector.java

```diff
@@ -124,25 +129,55 @@ void injectTargetAccountAndContainerFromBlobId(String blobIdStr, RestRequest res
+    Account targetAccount = (Account) restRequest.getArgs().get(RestUtils.InternalKeys.TARGET_ACCOUNT_KEY);
+    Container targetContainer = (Container) restRequest.getArgs().get(RestUtils.InternalKeys.TARGET_CONTAINER_KEY);
+    if (targetAccount == null || targetContainer == null) {
+      throw new RestServiceException("Account and container were not injected by BlobStorageService",
+          RestServiceErrorCode.InternalServerError);
+    } else if (targetAccount.equals(Account.UNKNOWN_ACCOUNT) && targetContainer.equals(
+        Container.UNKNOWN_CONTAINER)) {
+      String serviceId = blobInfo.getBlobProperties().getServiceId();
+      boolean isPrivate = blobInfo.getBlobProperties().isPrivate();
+      injectAccountAndContainerUsingServiceId(restRequest, serviceId, isPrivate);
+    }
+  }
+
+  /**
+   * Inject {@link Account} and {@link Container} into a {@link RestRequest} based on a blob's service ID and
+   * privacy setting.
    * @param restRequest The {@link RestRequest} to inject {@link Account} and {@link Container} object.
+   * @param serviceId The service ID associated with the blob.
+   * @param isPrivate The blob's privacy setting.
    * @throws RestServiceException if either of {@link Account} or {@link Container} object could not be found.
    */
-  private void injectAccountAndContainerUsingServiceId(RestRequest restRequest) throws RestServiceException {
-    String accountName = getHeader(restRequest.getArgs(), Headers.SERVICE_ID, false);
-    Account targetAccount = accountService.getAccountByName(accountName);
+  private void injectAccountAndContainerUsingServiceId(RestRequest restRequest, String serviceId, boolean isPrivate)
+      throws RestServiceException {
+    // First, try to see if a migrated account exists for the service ID.
+    Account targetAccount = accountService.getAccountByName(serviceId);
     if (targetAccount == null) {
       frontendMetrics.unrecognizedServiceIdCount.inc();
       logger.debug(
-          "Account cannot be found for accountName={} in put request with serviceId. Setting targetAccount to UNKNOWN_ACCOUNT",
-          accountName);
+          "Account cannot be found for put request with serviceId={}. Setting targetAccount to UNKNOWN_ACCOUNT",
+          serviceId);
+      // If a migrated account does not exist fall back to the UNKNOWN_ACCOUNT.
       targetAccount = Account.UNKNOWN_ACCOUNT;
     }
-    boolean isBlobPrivate = isPrivate(restRequest.getArgs());
+    // Either the UNKNOWN_ACCOUNT, or the migrated account should contain default public/private containers
     Container targetContainer = targetAccount.getContainerById(
-        isBlobPrivate ? Container.DEFAULT_PRIVATE_CONTAINER_ID : Container.DEFAULT_PUBLIC_CONTAINER_ID);
+        isPrivate ? Container.DEFAULT_PRIVATE_CONTAINER_ID : Container.DEFAULT_PUBLIC_CONTAINER_ID);
+    if (targetContainer == null) {
```

Comment:
This relies on an operational step: creating migrated service ID accounts with the right containers. If the correct containers were not created this could be null.

Reply:
This may not be possible right. either the account may not exist. If it exists, we will def have these two containers. Is my understanding right ?

Reply:
There is no logic in `updateAccounts()` that guarantees that those containers are present. We have made sure it is there for the migrated service ID accounts that we created, but it is not a given.